controllers/overcloudipset_controller.go

```diff
@@ -142,7 +143,22 @@ func (r *OvercloudIPSetReconciler) Reconcile(req ctrl.Request) (ctrl.Result, err
 
 			if reservationIP == "" {
 				// No reservation found, so create a new one
-				ip, reservation, err := common.AssignIP(*cidr, start, end, network.Status.Reservations, hostname)
+				// TODO: we generate the hostname/host IDKey in different controllers, which needs to be changed
+				//       when using Predictable IPs. The IDKey should remain when a node gets scaled down/deleted.
+				//       Its entry should not be removed from the IP lists and instead needs to be marked deleted.
+				//       For now set the role + count which will break things when removing a node in the middle.
+				ip, reservation, err := common.AssignIP(common.AssignIPDetails{
+					IPnet:               *cidr,
+					RangeStart:          start,
+					RangeEnd:            end,
+					Reservelist:         network.Status.Reservations,
+					ExcludeRanges:       []string{},
+					IDKey:               fmt.Sprintf("%s-%d", strings.ToLower(instance.Spec.Role), count),
+					Hostname:            hostname,
+					VIP:                 instance.Spec.VIP,
+					Role:                instance.Spec.Role,
+					AddToPredictableIPs: instance.Spec.AddToPredictableIPs,
+				})
```

Comment:
Can we just use `hostname` as the `IDKey` here?  The func that generates the hostnames uses `Role`+`-<lowest index available>`, and it will pick gaps in the middle when they're available (to address the concern in line 149).  I think I might be missing the larger picture, though, for why we're adding `IDKey`.

Reply:
Based on your comments in your changes to `pkg/common/hostname.go`, I think there is something I am not seeing that would provide the context I need to understand why the answer to my question above might be "no". :-)

Reply:
The idea of having IDKey is related to [1] to never reuse an IDKey when a host gets deleted. We need a way to always re/create ips-from-pool-all from scratch with the information of deleted hosts(marked as deleted):

```
Even if an overcloud node is deleted, its entry in the IP lists should not be removed. The IP list is based on the underlying Heat indices, which do not change even if nodes are deleted. To indicate that a given entry in the list is no longer used, the IP value can be replaced with a value such as “DELETED” or “UNUSED”.

In short, entries should never be removed from the IP lists, only changed or added.
```

Also having IDKey we have the chance, if we ever decide, to provide a way to set custom hostnames (compute-0 -> mysupercompute [2])and it won't conflict with creating ips-from-pool-all.

For the above we don't have an implementation in hostname.go yet, using IDKey here is just a pre-requisit. If we say
* hostnames are never being re-used
* index always grows, like we had cmp1-10, scaled down to cmp9, scaled up to 10 again, but the new 10th cmp get cmp11 as hostname 
* keep the old (removed) hostnames to track the status in ips-from-pool-all

then we could just switch to use hostnames. Hope that makes sense :)

[1] https://docs.openstack.org/project-deploy-guide/tripleo-docs/latest/provisioning/node_placement.html#predictable-ips

[2] https://docs.openstack.org/project-deploy-guide/tripleo-docs/latest/provisioning/node_placement.html#custom-hostnames

Reply:
I had the same thoughts as @abays with regards to IDKey (like could we just use hostname). 

And thanks @stuggi for the explanation. Perhaps we keep IdKey for now, but I wonder is there a way to hide it on the CLI or something to keep it more of an internal resource. It does look odd on the Status when you look at the description of an overcloudnet.